src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt

```diff
@@ -0,0 +1,396 @@
+        RsChangeSignatureProcessor(project, config.createChangeInfo())
+
+    override fun createReturnTypeCodeFragment(): PsiCodeFragment =
+        createTypeCodeFragment(myMethod.function, myMethod.function.retType?.typeReference)
+
+    override fun createCallerChooser(
+        title: String?,
+        treeToReuse: Tree?,
+        callback: Consumer<MutableSet<RsFunction>>?
+    ): CallerChooserBase<RsFunction>? = null
+
+    override fun validateAndCommitData(): String? = validateAndUpdateData()
+    override fun areButtonsValid(): Boolean = isValid
+
+    override fun updateSignature() {
+        updateState()
+        super.updateSignature()
+    }
+    override fun updateSignatureAlarmFired() {
+        super.updateSignatureAlarmFired()
+        validateButtons()
+    }
+
+    override fun canRun() {
+        val error = validateAndUpdateData()
+        if (error != null) {
+            throw ConfigurationException(error)
+        }
+
+        super.canRun()
+    }
+
+    /**
+     * Updates the config from UI elements that are not updated automatically and also the validity state of the dialog.
+     */
+    private fun updateState() {
+        isValid = validateAndUpdateData() == null
+    }
+
+    private fun validateAndUpdateData(): String? {
+        val factory = RsPsiFactory(config.function.project)
+
+        if (myNameField != null) {
+            val functionName = myNameField.text
+            if (validateName(functionName)) {
+                config.name = functionName
+            } else return "Function name must be a valid Rust identifier"
+        }
+
+        if (myReturnTypeField != null) {
+            val returnTypeText = myReturnTypeField.text
+            val returnType = if (returnTypeText.isBlank()) {
+                factory.createType("()")
+            } else {
+                (myReturnTypeCodeFragment as? RsTypeReferenceCodeFragment)?.typeReference
+            }
+            if (returnType != null) {
+                config.returnTypeDisplay = returnType
+            } else {
+                return "Function return type must be a valid Rust type"
+            }
+        }
+
+        val visField = visibilityComboBox
+        if (visField != null) {
+            if (visField.hasValidVisibility) {
+                config.visibility = visField.visibility
+            } else {
+                return "Function visibility must be a valid visibility specifier"
+            }
+        }
+
+        for ((index, parameter) in config.parameters.withIndex()) {
+            if (parameter.parsePat() == null) {
+                return "Parameter $index has invalid pattern"
+            }
+            if (parameter.type is ParameterType.Empty) {
+                return "Please enter type for parameter $index"
+            }
+            if (parameter.type is ParameterType.Invalid) {
+                return "Type entered for parameter $index is invalid"
+            }
```

Comment:
Wouldn't it be more convenient for user if we refer to parameter by name instead of by index?

Reply:
I wanted to do that, but the parameter might:
1) Not have a name (it might be empty)
2) It might be a complex pattern (e.g. `Struct { a: Struct2(x, y), b: Struct3(z), .. }`).

Therefore, I think that the index is more general. What do you think? Or maybe we can show the name if the pattern is non-empty and simple, and in other cases show the index.

Reply:
Oh, indeed, there are so many possibilities. Then let's keep index